tests/test_args_specs.py

```diff
@@ -0,0 +1,335 @@
+                    self.assertTrue('contents' in arg)
+                    self.assertTrue(isinstance(arg['contents'], dict))
+                    for path in arg['contents']:
+                        self.assertTrue(isinstance(path, str))
+                        if '(' in path or '[' in path:
+                            print(f"Warning: header {path} in {name}")
+                        self.validate(
+                            arg['contents'][path],
+                            f'{name}.contents.{path}',
+                            valid_types=valid_nested_types['directory'])
+                    attrs.remove('contents')
+
+            # iterate over the remaining attributes
+            # type-specific ones have been removed by this point
+            if 'name' in attrs:
+                self.assertTrue(isinstance(arg['name'], str))
+                attrs.remove('name')
+            if 'about' in attrs:
+                self.assertTrue(isinstance(arg['about'], str))
+                attrs.remove('about')
+            if 'required' in attrs:
+                # required value may be True, False, or a string that can be
+                # parsed as a python statement that evaluates to True or False
+                self.assertTrue(isinstance(arg['required'], bool) or
+                                isinstance(arg['required'], str))
+                attrs.remove('required')
+            if 'type' in attrs:
+                self.assertTrue(isinstance(arg['type'], str) or
+                                isinstance(arg['type'], set))
+                attrs.remove('type')
+            if 'regexp' in attrs:
+                # if it has 'regexp' it should have 'items' also
+                self.assertTrue('items' in arg)
+                self.assertTrue(isinstance(arg['regexp'], str))
+                re.compile(arg['regexp'])  # should compile without errors
+                attrs.remove('regexp')
+            if 'items' in attrs:
+                # if it has 'items' it should have 'regexp' also
+                self.assertTrue('regexp' in arg)
+                self.assertTrue(isinstance(arg['items'], str))
+                attrs.remove('items')
+            if 'validation_options' in attrs:
+                # the allowed validation_options properties are type-specific
+                opts = arg['validation_options']
+                if arg['type'] == 'csv':
+                    self.assertTrue(list(opts.keys()) == ['excel_ok'])
+                    self.assertTrue(isinstance(opts['excel_ok'], bool))
+                elif arg['type'] == 'number':
+                    self.assertTrue(list(opts.keys()) == ['expression'])
+                    self.assertTrue(isinstance(opts['expression'], str))
+                elif arg['type'] == 'freestyle_string':
+                    self.assertEqual(list(opts.keys()), ['regexp'])
+                    self.assertTrue(isinstance(opts['regexp'], dict))
```

Comment:
I see that sometimes the 'regexp' is expressed as a dict under 'validation_options', and sometimes it is expressed as a str under the 'fields' property of a vector/csv. That seems to differ by the type of the arg (freestyle_string vs vector/csv), but I wonder if it's necessary to have two different specifications for this. 

Like, should those csv/vector 'fields' also nest their 'regexp' or other validation options under a 'validation_options' attribute? Or should these 'freestyle_strings' have their 'expression' or 'regexp', etc elevated rather than nested inside 'validation_options'?

Reply:
I would like the `freestyle_string`s to have their `expression/regexp` elevated to the same level rather than nested inside `validation_options` (and I kind of think that should happen for most/all the other `validation_options` options too, see the end of [this comment](https://github.com/natcap/invest/pull/526#issuecomment-832195815)). Didn't want to make that change without first getting a consensus on it, though.

Reply:
As discussed below, we agreed to remove the whole `validation_options` thing and have all attributes at the same level. This is neat because now you can pass the whole arg dictionary into the type validation function e.g. `check_csv(csv_path, **arg_spec['csv_arg'])`